pkg/utils/kubernetes/networking/services.go

```diff
@@ -45,3 +48,92 @@ func WaitForServiceLoadBalancerAddress(ctx context.Context, c *kubernetes.Client
 		}
 	}
 }
+
+// WaitForServicePortConnected waits for a service provided by namespace/name
+// being able to connect by TCP. It creates a LoadBalancer with same selector and ports,
+// and try to establish TCP connection to the loadbalacer IP and port.
+func WaitForServicePortConnected(ctx context.Context, c *kubernetes.Clientset, namespace, name string, port int) error {
+	service, err := c.CoreV1().Services(namespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	lbServiceName := "templb-" + name
+	tempLoadBalancer := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      lbServiceName,
+			Labels: map[string]string{
+				"corresponding-service": name,
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Type: corev1.ServiceTypeLoadBalancer,
+			// copy the selector and ports of the service to check.
+			Selector: service.Spec.Selector,
+			Ports:    service.Spec.Ports,
+		},
+	}
+
+	// empty selector, we should create the endpoints separately.
+	if service.Spec.Selector == nil || len(service.Spec.Selector) == 0 {
+		endpoints, err := c.CoreV1().Endpoints(namespace).Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		// copy the endpoints for the lb service, and set metadata.
+		tempEndPioints := endpoints.DeepCopy()
+		tempEndPioints.ObjectMeta = metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      lbServiceName,
+			Labels: map[string]string{
+				"corresponding-service": name,
+			},
+		}
+		_, err = c.CoreV1().Endpoints(namespace).Create(ctx, tempEndPioints, metav1.CreateOptions{})
```

Comment:
I feel like I might be missing something simple, can you help me to better understand why we're _creating_ endpoints?

Reply:
some services may not have an effective `spec.selector`. Their upstreams are specified by an `endpoints` resource with the same name created separately (for example, service for validating webhook in KIC integration tests, which pointing to the webhook server run outside testing k8s cluster). So we should also create endpoints here with same configuration of target service. 
I explained this in the comments, please give more comments if they did not express well.